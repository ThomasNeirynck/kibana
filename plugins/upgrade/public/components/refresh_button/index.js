export { RefreshButton } from './refresh_button';
