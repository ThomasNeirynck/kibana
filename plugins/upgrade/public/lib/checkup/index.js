import {
  getFromApi,
} from '../request';


export async function getDeprecations() {
  return await getFromApi(`/api/migration/deprecations`);
}
