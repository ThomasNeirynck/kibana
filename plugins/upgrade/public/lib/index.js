export { timeout } from './util';
export {
  getAssistance,
  getMappingsAndSettings,
  createIndex,
  setReadOnly,
  runReindex,
  getTaskDetails,
  refreshIndex,
  verifyDocs,
  getSettingsAndAliases,
  updateRefreshInterval,
  replaceIndex,
  deleteTask,
  resetIndex,
  cancelTask,
  runUpgrade,
  isCompleted,
  isFailed,
  isNotStarted,
  isRunning,
  isStepCompleted,
  isStepFailed,
  isStepNotStarted,
  isStepRunning,
  isResettable,
  getStepMessage,
  wrapErrorMessage,
} from './reindex';
