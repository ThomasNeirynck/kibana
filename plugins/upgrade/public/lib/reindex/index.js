export {
  isCompleted,
  isFailed,
  isNotStarted,
  isRunning,
  isCanceled,
  isResettable,
  isCancelable,
  isStepCompleted,
  isStepFailed,
  isStepNotStarted,
  isStepRunning,
  isStepCanceled,
} from './state';

export {
  getStepMessage,
  wrapErrorMessage,
} from './messages';

export {
  getAssistance,
  getMappingsAndSettings,
  getSettingsAndAliases,
  updateRefreshInterval,
  createIndex,
  setReadOnly,
  runReindex,
  getTaskDetails,
  cancelTask,
  refreshIndex,
  verifyDocs,
  replaceIndex,
  deleteTask,
  resetIndex,
  runUpgrade,
} from './step_methods';

