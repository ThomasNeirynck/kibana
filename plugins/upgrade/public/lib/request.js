import chrome from 'ui/chrome';

function get(url) {
  return new Promise((resolve, reject) => {
    const req = new window.XMLHttpRequest();

    req.open('GET', url);

    req.onload = () => {
      if (req.status === 200) {
        resolve(JSON.parse(req.response));
      } else {
        reject(JSON.parse(req.response));
      }
    };

    req.onerror = () => {
      reject(new Error('Network error'));
    };

    req.send();
  });
}

export function getFromApi(url) {
  return get(chrome.addBasePath(url));
}

function post(url, data) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('POST', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send(JSON.stringify(data));
  });
}

export function postToApi(url, data) {
  return post(chrome.addBasePath(url), data);
}

function put(url, data) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('PUT', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send(JSON.stringify(data));
  });
}

export function putToApi(url, data) {
  return put(chrome.addBasePath(url), data);
}

function del(url) {
  return new Promise((resolve, reject) => {
    const xhr = new window.XMLHttpRequest();
    xhr.open('DELETE', url);
    xhr.setRequestHeader('Content-type', 'application/json');
    xhr.setRequestHeader('kbn-version', chrome.getXsrfToken());
    xhr.onload = () => {
      if (xhr.status === 200) {
        resolve(JSON.parse(xhr.response));
      } else {
        reject(JSON.parse(xhr.response));
      }
    };
    xhr.send();
  });
}

export function deleteFromApi(url) {
  return del(chrome.addBasePath(url));
}
