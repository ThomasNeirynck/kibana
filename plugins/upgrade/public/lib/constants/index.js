import { HomeView } from '../../components/home';
import { CheckupView } from '../../components/checkup';
import { ReindexView } from '../../components/reindex';
import { LoggingView } from '../../components/logging';


export {
  NUMBER_OF_STEPS,
  REINDEX_STEPS,
  UPGRADE_STEPS,
  STEP_RESULTS,
  INDEX_ACTION,
} from './reindex';

export {
  DEPRECATION_ISSUE_LEVELS,
} from './deprecations';

export const VIEWS = {
  HOME: {
    component: HomeView,
    label: 'Overview',
    location: 'home',
  },
  CHECKUP: {
    component: CheckupView,
    label: 'Cluster Checkup',
    location: 'checkup',
  },
  REINDEX: {
    component: ReindexView,
    label: 'Reindex Helper',
    location: 'reindex',
  },
  LOGGING: {
    component: LoggingView,
    label: 'Toggle Deprecation Logging',
    location: 'logging',
  }
};
export const DEFAULT_VIEW_ID = 'HOME';

export const HUMAN_READABLE_DELAY = 750;

export const LOADING_STATUS = {
  FORBIDDEN: 'FORBIDDEN',
  FAILURE: 'FAILURE',
  LOADING: 'LOADING',
  SUCCESS: 'SUCCESS',
  UNINITIALIZED: 'UNINITIALIZED',
};
