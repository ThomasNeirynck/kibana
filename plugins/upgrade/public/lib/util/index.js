import _ from 'lodash';
import chrome from 'ui/chrome';
import {
  KIBANA_INDEX_SUFFIX,
  INDEX_SUFFIX,
} from '../../../common/constants';

import { kibanaMappings } from '../reindex/kibana';


export function timeout(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export function getDisplayName(Component) {
  return Component.displayName || Component.name || 'Component';
}

export function getIndexSuffix(indexName) {
  const kibanaIndex = chrome.getInjected('kbnIndex');
  if (indexName === kibanaIndex) {
    return KIBANA_INDEX_SUFFIX;
  } else {
    return INDEX_SUFFIX;
  }
}

function remapStringFields(mappings) {
  return _.transform(mappings, (result, mapping, type) => {
    const newMapping = { ...mapping };

    newMapping.properties = _.transform(mapping.properties, (acc, properties, field) => {
      const newProperties = { ...properties };

      if (properties.type && properties.type === 'string') {
        if (properties.index && properties.index === 'not_analyzed') {
          newProperties.type = 'keyword';
          delete newProperties.fielddata;
        } else {
          newProperties.type = 'text';
        }

        if (properties.index && properties.index === 'no') {
          newProperties.index = false;
        } else {
          delete newProperties.index;
        }
      }

      acc[field] = newProperties;
    }, {});

    result[type] = newMapping;
  }, {});
}

export function getUpgradedMappings(indexName, mappings) {
  const kibanaIndex = chrome.getInjected('kbnIndex');

  return (indexName === kibanaIndex) ? kibanaMappings : remapStringFields(mappings);
}

export function getUpgradedSettings(settings) {
  return {
    ...(_.omit(settings, [
      'index.uuid',
      'index.creation_date',
      'index.version.created',
      'index.version.upgraded',
      'index.provided_name',
      'index.blocks',
      'index.legacy',
    ])),
    'index.refresh_interval': -1,
    'index.format': 6,
  };
}

export function getReindexBody(indexName) {
  const suffix = getIndexSuffix(indexName);
  const kibanaIndex = chrome.getInjected('kbnIndex');

  const body = {
    source: {
      index: `${ indexName }`,
    },
    dest: {
      index: `${ indexName }${ suffix }`,
    },
  };

  if (indexName === kibanaIndex) {
    body.script = {
      inline: `
        ctx._source = [ ctx._type : ctx._source ];
        ctx._source.type = ctx._type;
        ctx._id = ctx._type + ":" + ctx._id;
        ctx._type = "doc";
      `
    };
  }

  return body;
}

export function getActionsForAliasesBody(oldAliases, indexName) {
  const suffix = getIndexSuffix(indexName);
  const actions = _.transform(oldAliases, (result, aliasSettings, aliasName) => {
    const action = {
      add: {
        index: `${ indexName }${ suffix }`,
        alias: aliasName,
        ...aliasSettings,
      },
    };

    result.push(action);
  }, []);

  actions.push({
    remove_index: {
      index: indexName
    },
  });
  actions.push({
    add: {
      index: `${ indexName }${ suffix }`, "alias": indexName
    },
  });
  return actions;
}
