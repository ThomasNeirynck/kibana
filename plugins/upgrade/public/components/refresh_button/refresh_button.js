import React, { Component } from 'react';
import { timeout } from '../../lib';
import { HUMAN_READABLE_DELAY } from '../../lib/constants';
import {
  KuiButton,
  KuiButtonIcon,
} from 'ui_framework/components';

export class RefreshButton extends Component {
  state = {
    isRunning: false,
  };

  onClick = async () => {
    this.setState({ isRunning: true });

    try {
      await this.props.onClick();
      await timeout(HUMAN_READABLE_DELAY);
      this.setState({ isRunning: false });

    } catch (error) {
      this.setState({ isRunning: false });
    }
  }

  render() {
    const { buttonLabel, className } = this.props;
    const { isRunning } = this.state;

    return (
      <KuiButton
        buttonType="primary"
        className={ className }
        icon={ <KuiButtonIcon className="fa-refresh" /> }
        disabled={ isRunning }
        onClick={ this.onClick }
      >
        { isRunning ? 'Running...' : buttonLabel }
      </KuiButton>
    );
  }
}

RefreshButton.propTypes = {
  buttonLabel: React.PropTypes.string,
  className: React.PropTypes.string,
  onClick: React.PropTypes.func,
};

RefreshButton.defaultProps = {
  buttonLabel: 'Run',
  className: null,
  onClick: () => {},
};

