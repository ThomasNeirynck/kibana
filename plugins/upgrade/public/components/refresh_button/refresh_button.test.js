import { shallow } from 'enzyme';
import React from 'react';

import { RefreshButton } from './refresh_button';
import { timeout } from '../../lib/util';

jest.mock('ui/chrome', () => {}, { virtual: true });
jest.mock('ui/notify/notifier', () => ({
  Notifier: function Notifier() {},
}), { virtual: true });
jest.mock('../../lib/util');


describe('RefreshButton', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('renders the className property', () => {
    const component = (
      <RefreshButton className="specificStatusGroup" />
    );

    expect(shallow(component)).toMatchSnapshot();
  });

  test('renders the buttonLabel property', () => {
    const component = (
      <RefreshButton buttonLabel="A different label" />
    );

    expect(shallow(component)).toMatchSnapshot();
  });

  test('is disabled until the promises returned by onClick and timeout have been resolved', async () => {
    let resolveOnClick;
    const onClickPromise = new Promise((resolve) => resolveOnClick = resolve);
    const onClick = jest.fn().mockReturnValue(onClickPromise);
    let resolveTimeout;
    const timeoutPromise = new Promise((resolve) => resolveTimeout = resolve);
    timeout.mockReturnValue(timeoutPromise);
    const component = (
      <RefreshButton onClick={ onClick } />
    );

    const wrapper = shallow(component);

    expect(wrapper.find('KuiButton').prop('disabled')).toBe(false);
    wrapper.find('KuiButton').prop('onClick')();

    expect(wrapper.find('KuiButton').prop('disabled')).toBe(true);
    expect(wrapper.find('KuiButton').contains('Running...')).toBe(true);

    resolveOnClick();
    await expect(onClickPromise).resolves.toBe();

    expect(wrapper.find('KuiButton').prop('disabled')).toBe(true);
    expect(wrapper.find('KuiButton').contains('Running...')).toBe(true);

    resolveTimeout();
    await expect(timeoutPromise).resolves.toBe();

    expect(timeout).toHaveBeenCalledWith(expect.any(Number));
    expect(wrapper.find('KuiButton').prop('disabled')).toBe(false);
    expect(wrapper.find('KuiButton').contains('Running...')).toBe(false);
  });

  test('is disabled until the promise returned by onClick has been rejected', async () => {
    let rejectOnClick;
    const onClickPromise = new Promise((resolve, reject) => rejectOnClick = reject);
    const onClick = jest.fn().mockReturnValue(onClickPromise);
    const component = (
      <RefreshButton onClick={ onClick } />
    );

    const wrapper = shallow(component);

    expect(wrapper.find('KuiButton').prop('disabled')).toBe(false);
    wrapper.find('KuiButton').prop('onClick')();

    expect(wrapper.find('KuiButton').prop('disabled')).toBe(true);
    expect(wrapper.find('KuiButton').contains('Running...')).toBe(true);

    rejectOnClick();
    await expect(onClickPromise).rejects.toBe();

    expect(timeout).not.toHaveBeenCalled();
    expect(wrapper.find('KuiButton').prop('disabled')).toBe(false);
    expect(wrapper.find('KuiButton').contains('Running...')).toBe(false);
  });
});
