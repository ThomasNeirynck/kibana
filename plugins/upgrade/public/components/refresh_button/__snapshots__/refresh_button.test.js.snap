// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefreshButton renders the buttonLabel property 1`] = `
<KuiButton
  buttonType="primary"
  className={null}
  disabled={false}
  icon={
    <KuiButtonIcon
      className="fa-refresh"
    />
  }
  onClick={[Function]}
>
  A different label
</KuiButton>
`;

exports[`RefreshButton renders the className property 1`] = `
<KuiButton
  buttonType="primary"
  className="specificStatusGroup"
  disabled={false}
  icon={
    <KuiButtonIcon
      className="fa-refresh"
    />
  }
  onClick={[Function]}
>
  Run
</KuiButton>
`;
